src/main/java/io/confluent/connect/jdbc/sink/metadata/FieldsMetadata.java

```diff
@@ -129,7 +132,33 @@ public static FieldsMetadata extract(
       );
     }
 
-    return new FieldsMetadata(keyFieldNames, nonKeyFieldNames, allFields);
+    final Map<String, SinkRecordField> allFieldsOrdered = new LinkedHashMap<>();
+    for (String fieldName : JdbcSinkConfig.DEFAULT_KAFKA_PK_NAMES) {
+      if (allFields.containsKey(fieldName)) {
+        allFieldsOrdered.put(fieldName, allFields.get(fieldName));
+      }
+    }
+
+    if (valueSchema != null) {
+      for (Field field : valueSchema.fields()) {
+        String fieldName = field.name();
+        if (allFields.containsKey(fieldName)) {
+          allFieldsOrdered.put(fieldName, allFields.get(fieldName));
+        }
+      }
+    }
+
+    if (allFieldsOrdered.size() < allFields.size()) {
+      ArrayList<String> fieldKeys = new ArrayList<>(allFields.keySet());
+      Collections.sort(fieldKeys);
+      for (String fieldName : fieldKeys) {
+        if (!allFieldsOrdered.containsKey(fieldName)) {
+          allFieldsOrdered.put(fieldName, allFields.get(fieldName));
+        }
+      }
+    }
```

Comment:
If there's fields missing (possible if pkMode != RECORD_VALUE and valueSchema = NULL) then add the remaining fields sorted by name for deterministic ordering.